Sming/Arch/Esp32/Core/HardwarePWM.cpp

```diff
@@ -0,0 +1,305 @@
+	uint32_t pwm_duty_init[SOC_LEDC_CHANNEL_NUM]; // pwm duty
+	for(uint8_t i = 0; i < no_of_pins; i++) {
+		pwm_duty_init[i] = 0; // Start with zero output
+		channels[i] = pins[i];
+
+		/* 
+		 *  Prepare and then apply the LEDC PWM timer configuration.
+		 *  This may cofigure the same timer more than once (in fact up to 8 times)
+		 *  which should not be an issue, though, since the values should be the same for all timers
+		 */
+		// The two groups (if available) are operating in different speed modes, hence speed mode is an alias for group or vice versa
+		ledc_timer_config_t ledc_timer{
+			.speed_mode = pinToGroup(i),
+			.duty_resolution = LEDC_TIMER_10_BIT, // todo: make configurable later
+			.timer_num = pinToTimer(i),
+			.freq_hz = periodToFrequency(DEFAULT_PERIOD), // todo: make configurable later
+			.clk_cfg = LEDC_AUTO_CLK,
+		};
+
+		debug_d("ledc_timer.\r\n"
+				"\tspeed_mode: %i\r\n"
+				"\ttimer_num: %i\r\n"
+				"\tduty_resolution: %i\r\n"
+				"\tfreq: %i\n\tclk_cfg: %i\r\n\n",
+				ledc_timer.speed_mode, ledc_timer.timer_num, ledc_timer.duty_resolution, ledc_timer.freq_hz,
+				ledc_timer.clk_cfg);
+		ESP_ERROR_CHECK(ledc_timer_config(&ledc_timer));
+
+		/*
+		 * Prepare and then apply the LEDC PWM channel configuration
+		 */
+		ledc_channel_config_t ledc_channel{
+			.gpio_num = pins[i],
+			.speed_mode = pinToGroup(i),
+			.channel = pinToChannel(i),
+			.intr_type = LEDC_INTR_DISABLE,
+			.timer_sel = pinToTimer(i),
+			.duty = 0,
+			.hpoint = 0,
+		};
+		debug_d("ledc_channel\n"
+				"\tspeed_mode: %i\r\n"
+				"\tchannel: %i\r\n"
+				"\ttimer_sel %i\r\n"
+				"\tintr_type: %i\r\n"
+				"\tgpio_num: %i\r\n"
+				"\tduty: %i\r\n"
+				"\thpoint: %i\r\n\n",
+				pinToGroup(i), pinToChannel(i), pinToTimer(i), ledc_channel.intr_type, pins[i], 0, 0);
+		ESP_ERROR_CHECK(ledc_channel_config(&ledc_channel));
+		ledc_bind_channel_timer(pinToGroup(i), pinToChannel(i), pinToTimer(i));
+	}
+	maxduty = maxDuty(DEFAULT_RESOLUTION);
+}
+
+HardwarePWM::~HardwarePWM()
+{
+	// Stop pwm for all pins and set idle level to 0
+	for(uint8_t i = 0; i < channel_count; i++) {
+		ledc_stop(pinToGroup(i), pinToChannel(i), 0);
+	}
+}
+
+uint8_t HardwarePWM::getChannel(uint8_t pin)
+{
+	for(uint8_t i = 0; i < channel_count; i++) {
+		if(channels[i] == pin) {
+			// debug_d("getChannel %d is %d", pin, i);
+			return i;
+		}
+	}
+	return -1;
+}
+
+uint32_t HardwarePWM::getDutyChan(uint8_t chan)
+{
+	// esp32 defines the frequency / period per timer
+	return (chan == PWM_BAD_CHANNEL) ? 0 : ledc_get_duty(pinToGroup(chan), pinToChannel(chan));
+}
+
+bool HardwarePWM::setDutyChan(uint8_t chan, uint32_t duty, bool update)
+{
+	if(chan == PWM_BAD_CHANNEL) {
+		return false;
+	}
+
+	if(duty <= maxduty) {
+		ESP_ERROR_CHECK(ledc_set_duty(pinToGroup(chan), pinToChannel(chan), duty));
+		if(update) {
+			ESP_ERROR_CHECK(ledc_update_duty(pinToGroup(chan), pinToChannel(chan)));
+			//update();
+		}
```

Comment:
If the `update` method isn't implemented then this needs changing so it ignores the `update` flag and (at least for now) *always* updates the channel.